db/src/db.rs

```diff
@@ -416,45 +437,438 @@ impl DB {
+
+                    let added = true;
+                    if attribute.multival {
+                        non_fts_many.push((e, a, typed_value, added));
+                    } else {
+                        non_fts_one.push((e, a, typed_value, added));
+                    }
+                    Ok(())
+                },
+
+                Entity::Retract {
+                    e: entmod::EntidOrLookupRef::Entid(ref e_),
+                    a: ref a_,
+                    v: entmod::ValueOrLookupRef::Value(ref v_) } => {
+
+                    let e: i64 = match e_ {
+                        &entmod::Entid::Entid(ref e__) => *e__,
+                        &entmod::Entid::Ident(ref e__) => *self.schema.require_entid(&e__.to_string())?,
+                    };
+
+                    let a: i64 = match a_ {
+                        &entmod::Entid::Entid(ref a__) => *a__,
+                        &entmod::Entid::Ident(ref a__) => *self.schema.require_entid(&a__.to_string())?,
+                    };
+
                     let attribute: &Attribute = self.schema.require_attribute_for_entid(&a)?;
+                    if attribute.fulltext {
+                        bail!(ErrorKind::NotYetImplemented(format!("Transacting :db/fulltext entities is not yet implemented: {:?}", entity)))
+                    }
 
                     // This is our chance to do schema-aware typechecking: to either assert that the
                     // given value is in the attribute's value set, or (in limited cases) to coerce
                     // the value into the attribute's value set.
                     let typed_value: TypedValue = self.to_typed_value(v_, &attribute)?;
 
-                    // Now we can represent the typed value as an SQL value.
-                    let (value, value_type_tag): (ToSqlOutput, i32) = typed_value.to_sql_value_pair();
+                    let added = false;
 
-                    // Fun times, type signatures.
-                    let values: [&ToSql; 9] = [&e, &a, &value, &tx, &value_type_tag, &attribute.index, to_bool_ref(attribute.value_type == ValueType::Ref), &attribute.fulltext, &attribute.unique_value];
-                    stmt.insert(&values[..])?;
+                    if attribute.multival {
+                        non_fts_many.push((e, a, typed_value, added));
+                    } else {
+                        non_fts_one.push((e, a, typed_value, added));
+                    }
                     Ok(())
                 },
-                // TODO: find a better error type for this.
-                _ => panic!(format!("Transacting entity not yet supported: {:?}", entity))
+
+                _ => bail!(ErrorKind::NotYetImplemented(format!("Transacting this entity is not yet implemented: {:?}", entity)))
             }
         }).collect();
 
-        let x: Result<Vec<()>> = r.into_iter().collect();
-        x.map(|_| ())
+        let r: Result<Vec<()>> = r.into_iter().collect();
+        r?;
```

Comment:
Sadly, the "simplification" is the turbofish (`::<>`) which is even harder to read.